Use Arc features in Hibernate extensions for eager startup and active/inactive

yrodiere opened the pull request:
Draft for now.

To do:

- [x] Implement basic functionality for Hibernate ORM/Reactive
- [x] Run basic tests locally for Hibernate ORM/Reactive
- [x] Handle FIXME in the code about initialization (one-by-one vs. parallel threads)
- [x] Handle FIXME in the code about Panache
- [x] Run the whole test suite and not just a few tests in the ORM extension
- [x] Handle Hibernate Envers
- [x] Handle Hibernate Search
- [x] ~~Handle Hibernate Validator ([remove `JPATraversableResolver` if ORM deactivated?](https://github.com/quarkusio/quarkus/blob/28a73f6652d1129cd78f677256060ccee16f367d/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorRecorder.java#L299))~~ => Bad idea, this would be complex to implement, and also to use (if entity types are used outside of PUs, why should the PU active/inactive status matter?
- [x] Update documentation
- [x] Write migration guide entry

* Closes #43594 

Migration guide entry:

`````asciidoc
== Hibernate ORM

[[hibernate-orm-inactive-fail-fast]]
=== Hibernate ORM usage fails fast if persistence unit is deactivated

In previous versions of Quarkus, when the persistence unit was deactivated (`quarkus.hibernate-orm.active=false`), the application would be allowed to start successfully, but could fail much later upon first access to a Hibernate bean (`EntityManager`, `Session`, `SessionFactory`, ...).

Starting with this version of Quarkus, persistence units that are deactivated will lead to a startup failure if Quarkus can detect they are used:

* Static CDI injection points involving the persistence unit (`@Inject Session session`, `@Inject SessionFactory sessionFactory`, ...) will cause application startup to fail with an explicit, actionable message.
* Dynamic retrieval of the persistence unit (e.g. through `CDI.getBeanContainer()`/`Arc.instance()`, or by injecting an `Instance<Session>`) will not be detected on startup, but will cause an exception to be thrown with an explicit, actionable message.

The <<hibernate-reactive-inactive-fail-fast,Hibernate Reactive>> and <<hibernate-search-inactive-fail-fast,Hibernate Search>> extensions implement a similar behavior for their `Mutiny.SessionFactory`/`SearchSession`/`SearchMapping` CDI beans.

If your application needs to inject a `EntityManager/`Session`/`EntityManagerFactory`/`SessionFactory` or related bean for a persistence unit that can potentially be deactivated or have no URL, you may encounter a startup failure similar to this:

```
io.quarkus.arc.InactiveBeanException: Bean is not active: SYNTHETIC bean [class=org.hibernate.Session, id=QdxvhYUzh5cgfepxgL43K9-OFvU]
Reason: Persistence unit '<default>' was deactivated through configuration properties. To activate the persistence unit, set configuration property 'quarkus.hibernate-orm.active' to 'true' and configure datasource '<default>'. Refer to https://quarkus.io/guides/datasource for guidance.
To avoid this exception while keeping the bean inactive:
	- Configure all extensions consuming this bean as inactive as well, if they allow it, e.g. 'quarkus.someextension.active=false'
	- Make sure that custom code only accesses this bean if it is active
	- Inject the bean with 'Instance<org.hibernate.Session>' instead of 'org.hibernate.Session'
This bean is injected into:
	- io.quarkus.hibernate.orm.config.ConfigActiveFalseStaticInjectionTest$MyBean#session
[...]
```

To avoid this failure, follow the instructions in the exception message.
In particular, consider injecting the problematic bean as an `InjectableInstance` (for example `@Inject InjectableInstance<Session> session`) instead.
You can check whether that bean is active using `.getHandle().getBean().isActive().result()` on the `InjectableInstance`, and retrieve the bean instance using `InjectableInstance#get()`.
See https://quarkus.io/guides/hibernate-orm#persistence-unit-active for an example of such a setup for `Session`.

== Hibernate Reactive

[[hibernate-reactive-inactive-fail-fast]]
=== Hibernate Reactive usage fails fast if persistence unit is deactivated

Starting with this version of Quarkus, Hibernate Reactive persistence units that are deactivated will lead to a startup failure if Quarkus can detect they are used.

See the <<hibernate-orm-inactive-fail-fast,datasources,similar migration guide entry for Hibernate ORM>> for more information.

== Hibernate Search

[[hibernate-search-inactive-fail-fast]]
=== Hibernate Search usage fails fast if persistence unit is deactivated

Starting with this version of Quarkus, Hibernate Search mappings that are deactivated will lead to a startup failure if Quarkus can detect they are used.

See the <<hibernate-orm-inactive-fail-fast,datasources,similar migration guide entry for Hibernate ORM>> for more information.
`````

yrodiere committed:
Refactor tests around active/inactive for Hibernate ORM/Reactive/Search

So that we can more easily see what the next commit changes.

yrodiere committed:
Document and test that persistence units can be deactivated with Hibernate Reactive too

yrodiere committed:
Test that Hibernate Envers and Search don't fail to start when Hibernate ORM is deactivated

yrodiere committed:
Use Arc features in Hibernate extensions for eager startup and active/inactive

marko-bekhta reviewed (commented):
Thanks! Looks nice 😃 🎉 

I had some questions, but feel free to ignore those 😄

yrodiere committed:
Clarify that injecting `Instance<MyBean>` won't fail startup if MyBean is inactive

Only actually attempting to retrieve the bean instance will trigger a
failure.

Co-Authored-By: marko-bekhta <marko.prykladna@gmail.com>

yrodiere committed:
Fix typo in Hibernate ORM documentation example

yrodiere reviewed (commented):
I addressed your comments @marko-bekhta , thank you!

Sanne commented:
> We're good to go, I just need an approval from someone with push access... @Sanne would you please do that?

Done. Let's propose expandanding the push access team with them both?

yrodiere commented:
Thank you.

> Done. Let's propose expanding the push access team with them both?

I'll try again. Last time I did it didn't go well due to missing official processes for that and/or the Commonhaus transfer. Let's hope that's settled.
